Refactor out channel groups

jrainville opened the pull request:
Fixes https://github.com/status-im/status-desktop/issues/12595
Status-go PR: https://github.com/status-im/status-go/pull/4997

This PR gets rid of the concept of channelGroups.

Instead, we use the communities and chats directly, to keep the same concepts as status-go.

This PR affects a lot of places:
- sections
- chats
- members
- settings (notifications, communities)
- app search
- edits to any sections or chats
- canPost values (especially reactions)

It's a lot of changes, but in the end, it's more deletion than additions, and it also should make the code more straightforward (no more conversion from community to channelGroup for example).

The only thing I wasn't able to fully address is the fact that CommunityChats still in theory contain more info than "normal" Chats, but they all still convert to ChatDto, so we are still at risk of missing fields sometimes. It should all be covered here, but it's something we can keep improving. The problem comes from status-go. The Chats store contains all chats, including community chats, but not the CommunityChat properties.

[channelGroup-refactor.webm](https://github.com/status-im/status-desktop/assets/11926403/8aea3993-3eac-4cd6-93cd-4bb5969f3287)

jrainville commented:
I'll leave the commits as is so you guys can see the evolution. I'll squash when merging

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 8bd97d21132c668d3944ee9940df4cbad136ec86 | [#1](https://ci.status.im/job/status-desktop/job/prs/job/macos/job/aarch64/job/package/job/PR-14202/1/) | 2024-03-27 21:03:30 | ~7 min | `macos/aarch64` | [:apple:`dmg`](https://status-im-desktop-prs.ams3.cdn.digitaloceanspaces.com/StatusIm-Desktop-240327-205629-8bd97d21132c668d3944ee9940df4cbad136ec86-pr14202-aarch64.dmg) |

osmaczko commented on src/app/modules/main/chat_section/controller.nim:
```diff
@@ -427,9 +427,24 @@ proc getAllChats*(self: Controller, communityId: string): seq[ChatDto] =
+    community = self.getMyCommunity()
+    # chats = community.chats
+    chats = self.chatService.getChatsForCommunity(community.id)
+    echo "name ", community.name
+    echo "CHATS LEN ", chats.len
```
echo

osmaczko commented on src/app/modules/main/chat_section/controller.nim:
```diff
@@ -427,9 +427,24 @@ proc getAllChats*(self: Controller, communityId: string): seq[ChatDto] =
+    echo "name ", community.name
+    echo "CHATS LEN ", chats.len
+    # Fetch community chat members as we lazy load them
+    # TODO add fetch
```
Should be resolved before merge?

jrainville replied:
I the end I didn't need to do that. I forgot that we have the members in the community chats already. I just removed the TODO.

osmaczko commented on src/app/modules/main/chat_section/module.nim:
```diff
@@ -690,6 +692,16 @@ proc addNewChat(
+  var hideIfPermissionsNotMet = false
+  var viewersCanPostReactions = true
+  if self.controller.isCommunity:
+    let communityChat = community.getSpecificCommunityChat(chatDto.id)
```
Not sure I understand this part, isn't `communityChat` the same type (`ChatDto`) as `chatDto` that is already processed? Aren't we basically refetching the same `chatDto` here?

jrainville replied:
Same thing as explained above. The CommunityChat contains more info still. It will be addressed in a further refactor.

jrainville commented on src/app/modules/main/app_search/module.nim:
```diff
@@ -146,80 +162,85 @@ method searchMessages*(self: Module, searchTerm: string) =
-            if isCommunity:
-                SEARCH_RESULT_CHANNELS_SECTION_NAME
-              else:
-                SEARCH_RESULT_CHATS_SECTION_NAME,
```
Nice catch. I removed all the code that used `isCommunity` but missed the difference here. It was not a huge deal since it was just a string difference, but I fixed it.

jrainville commented on src/app/modules/main/chat_section/chat_content/users/controller.nim:
```diff
@@ -115,11 +115,18 @@ proc init*(self: Controller) =
 
 proc getChatMembers*(self: Controller): seq[ChatMember] =
-  return self.chatService.getChatById(self.chatId).members
+  if self.belongsToCommunity:
+    let myCommunity = self.getMyCommunity()
+    return myCommunity.getSpecificCommunityChat(self.chatId).members
+  else:
```
Sadly no. That's why I tried to explain in the description. The chats cache in the chatService contains all the active chats coming from the AllChats store from status-go.

That AllChats store contains all known chats, including the community chats, however, it uses the `protocol.Chat` object type, whereas chats that come from the Community are `CommunityChat`s.

`CommunityChat` contains less info, but it has some extra fields that only apply to communities, like `CanPostReactions`. Same thing for members, the AllChats store doesn't populate the Chat members for communities, only for GroupChats.

Both those types of Chats get converted to ChatDto on Desktop's side, so it's obviously confusing.

I'll open a new issue to document those differences and do another refactor later. The goal will be to make sure that our chat cache is usable at all times and up to date will all data, even from a community.

jrainville replied:
> Is mentioned above API something we shouldn't use because it misses CommunityChat details?

Indeed

jrainville replied:
Issue opened here: https://github.com/status-im/status-desktop/issues/14219

jrainville commented on src/app_service/service/chat/service.nim:
```diff
@@ -187,126 +177,62 @@ QtObject:
     result.unviewedMessagesCount = 0
 
-    for chat in self.channelGroups[communityId].chats:
+    let myPubKey = singletonInstance.userProfile.getPubKey()
+    var commId = communityId
+    if communityId == myPubKey:
+      commId = ""
```
We don,t want to return. This function is used by the personal chat section too. I'll rename the param to `sectionId` id, because I agree that using `communityId` is confusing.
I'll also add a comment to explain the conversion

jrainville commented on src/app_service/service/chat/service.nim:
```diff
@@ -842,7 +692,8 @@ QtObject:
       let checkChannelPermissionsResponse = rpcResponseObj["response"]["result"].toCheckChannelPermissionsResponseDto()
-      self.channelGroups[communityId].channelPermissions.channels[chatId] = checkChannelPermissionsResponse
+
+      # TODO need to save permissions somewhere?
```
Ah, I did check and it seems to work fine still. We can use `channelPermissions` from the community directly.
I'll remove the TODO, it's a leftover

jrainville commented on src/app_service/service/community/service.nim:
```diff
@@ -2465,3 +2451,18 @@ QtObject:
     except Exception as e:
       error "error promoting self to control node", msg = e.msg
+
+  proc chatsWithCategoryHaveUnreadMessages*(self: Service, communityId: string, categoryId: string): bool =
```
I didn't choose that name FYI :laughing: 
I'll rename like you said, it's way better

jrainville reviewed (commented):
Thanks for the comments. Nice finds on some of these, I'll address

jrainville commented:
@osmaczko addressed your comments

anastasiyaig commented:
@jrainville tests are failing because they are not able to click emoji selector in account popups, did your changes impact that or its my tests? The emoji button is clicked it seems but no emoji selector showed up

![image](https://github.com/status-im/status-desktop/assets/82375995/7f84b73b-4a5c-472c-ac0d-cc5b87004323)

jrainville commented:
> @jrainville tests are failing because they are not able to click emoji selector in account popups, did your changes impact that or its my tests? The emoji button is clicked it seems but no emoji selector showed up

I didn't touch the QML at all. Let me try locally to see if I could have had an impact

anastasiyaig commented:
alright i will check!

anastasiyaig commented:
@jrainville i downloaded this build from that PR to my 5.15.2 linux machine and noticed that on the first launch, there is an error on the chat screen. Then when i access the wallet for example, chat button is gone at all. And i can reproduce the issue with emoji popup

![Screenshot from 2024-03-29 09-38-12](https://github.com/status-im/status-desktop/assets/82375995/21df0093-b0b8-4c0a-8281-02d4393a5a02)
![Screenshot from 2024-03-29 09-38-34](https://github.com/status-im/status-desktop/assets/82375995/09662631-eeea-4708-9d18-4c1451848d15)

[Screencast from 29-03-24 09:38:43.webm](https://github.com/status-im/status-desktop/assets/82375995/cc24480b-5d12-4184-a936-1a9286c59e22)


```
Critical: AppMain: Unknown section type (qrc:/app/mainui/AppMain.qml:1168, expression for currentIndex)
Warning: qrc:/app/mainui/AppMain.qml:1138:21: Unable to assign [undefined] to int (qrc:/app/mainui/AppMain.qml:1138, )
Warning: qrc:/app/AppLayouts/Onboarding/OnboardingLayout.qml:149:9: Unable to assign [undefined] to QObject* (qrc:/app/AppLayouts/Onboarding/OnboardingLayout.qml:149, )
Warning: qrc:/main.qml:173:9: Unable to assign [undefined] to QObject* (qrc:/main.qml:173, )
WRN 2024-03-29 09:37:16.126+03:00 Error decoding signal                      topics="signals-manager" tid=65150 file=signals_manager.nim:49 err="Unknown signal received: mediaserver.started"
ERR 2024-03-29 09:37:16.126+03:00 error get active chats:                    topics="chat-service" tid=65150 file=service.nim:212 errDesription="No chats returned"
```

jrainville commented:
@anastasiyaig nice catch. I forgot to test a new account. I only tested big accounts.

The problem was pretty simple, I was returning an error when the function didn't return any chats, but that's normal for new accounts, so no need to return an error in that case.

jrainville commented:
Tests work now.

@anastasiyaig when you or one of your QAs have time, it would be good to test this PR out.
Of course, 2.28 issues take priority, so whenever.

anastasiyaig commented:
I ran all autotests we have and there is only 1 failure when checking online / offline indicator https://ci.status.im/job/status-desktop/job/e2e/job/manual/1728/allure/#suites/c779a5502ab37d502705212e4949e67a/91ece6f15123e01c/ but i think it is just the way how online indicator works. sometimes it shows offline for online user and so on
@glitchminer ^^

glitchminer commented:
Hey @jrainville , are these related to this change?

Channels and categories can be created ok but members are not seeing the same arrangement.

1. Create community, category and channels
2. Add a member
![image](https://github.com/status-im/status-desktop/assets/50769329/4b13c537-0443-45e0-8b96-d36d8f5291ca)

---
Noted that the permissions overlay is being shown when visiting an open community for the first time, even when no permissions are set
![image](https://github.com/status-im/status-desktop/assets/50769329/23003358-aaf4-4ab5-9638-ae305ebeedc5)

jrainville commented:
> are these related to this change?

Ah it's possible! Nice finds. I'll fix

jrainville commented:
@glitchminer the issues were indeed in my PR. They are fixed now.

For those curious, the problem was again the difference between CommunityChat and Chat. I noted the refactor in the fix I did.

There was also an issue I discovered which was that if you spectated a community without joining, when you restated, the community didn't appear and you couldn't see it anymore. I fixed that too.

jrainville commented:
I rebased and fixed conflicts. It builds, but I couldn't test it because it won,t let me start the app because of the changes in the status-go build system.

Hopefully it gets addressed soon.

In theory, the CI should still build, so @glitchminer you should be able to continue testing

jrainville commented:
Ok, now it's rebased on top of Jakub's fix. I was able to test that the basic functionality still works. The CI should work now too

jrainville commented:
@anastasiyaig @glitchminer will you have time to re-check this PR this week? If not, maybe it's worth merging and addressing issues on master? 

I'm asking, because it has been opened fro more than a month now and it's better to have it merged not too close to the next release, that way we can find issues faster and not on an RC.

anastasiyaig commented:
lets try to get it tested tomorrow, if @glitchminer has capacity. if not, then lets just test from master after merge @jrainville
